extensions/data-plane/data-plane-spi/src/main/java/org/eclipse/dataspaceconnector/dataplane/spi/pipeline/ParallelSink.java

```diff
@@ -0,0 +1,102 @@
+/*
+ *  Copyright (c) 2021 Microsoft Corporation
+ *
+ *  This program and the accompanying materials are made available under the
+ *  terms of the Apache License, Version 2.0 which is available at
+ *  https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Contributors:
+ *       Microsoft Corporation - initial API and implementation
+ *
+ */
+package org.eclipse.dataspaceconnector.dataplane.spi.pipeline;
+
+import org.eclipse.dataspaceconnector.common.stream.PartitionIterator;
+import org.eclipse.dataspaceconnector.dataplane.spi.result.TransferResult;
+import org.eclipse.dataspaceconnector.spi.monitor.Monitor;
+import org.eclipse.dataspaceconnector.spi.result.AbstractResult;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+
+import static java.util.concurrent.CompletableFuture.supplyAsync;
+import static java.util.stream.Collectors.toList;
+import static org.eclipse.dataspaceconnector.common.async.AsyncUtils.asyncAllOf;
+import static org.eclipse.dataspaceconnector.spi.response.ResponseStatus.ERROR_RETRY;
+
+/**
+ * Writes data in parallel.
+ */
+public abstract class ParallelSink implements DataSink {
+    protected String requestId;
+    protected int partitionSize = 5;
+    protected ExecutorService executorService;
+    protected Monitor monitor;
+
+    @Override
+    public CompletableFuture<TransferResult> transfer(DataSource source) {
```

Comment:
Shouldn't we have test cases for this class even though subclasses may have them?

Reply:
Yes, I created an item in our backlog already. However I believe this is a separate item from this PR (here I only factored out existing code from HttpDataSink)

Reply:
Prior, HttpDataSink tests covered that. ParalellSink is now refactored out in this PR so it needs to have test coverage before this is merged.

Reply:
HttpDataSink extends this new class, that just factors out a method from it, so coverage is not changed. ParallelSink has 100% coverage. Is your concern specifically that it should have coverage from tests *within its own module?* I can look into that.